### DISCUSSION TITLE
WA DL REGEX

### DISCUSSION THREAD

**Discussion by ammills01** [microsoft/presidio#388] [General]
The current WA DL REGEX matches any 12 letters, which can have see quite a few false positives. I called the Washington Department of Licensing today and she mentioned the first three characters always start with _WDL_ then the characters after are a random set of alphanumeric characters for a total of 12. I asked if there was documentation on how those were generated to build a better pattern and she put me on hold to ask her support folks. When she came back on the phone she seemed real skeptical of me and directed me to the PR department. I've emailed them, but it's the government so I hope for a response at some point in the next 12 months... Anyway, I was wondering if anyone else here had insight into how those WA DL numbers are made? The source the US DL recognizer lists, https://ntsi.com/drivers-license-format, does not seem to be updated with the schema the lady mentioned on the phone. If I can find official documentation I'll update the REGEX and send in a pull request.

It's worth noting she said the was the old schema: http://www.highprogrammer.com/alan/numbers/dl_us_wa.html

**Comment by will-saunders** [comment]
Washington state changed its methodology for assigning drivers license numbers in 2018.  The old system was based on the holder's name and year of birth, and had become a privacy risk.  Some of the old licenses are still in circulation, though with REALID the numbers are diminishing fast.  Here are the new and old patterns with comments.  My regex syntax may well be flawed, so please read the comments.

Item | Pattern | Comments
-- | -- | --
Current Washington Driver’s License | WDL[A-Z0-9][A-Z0-9][A-Z0-9][A-Z0-9][A-Z0-9][A-Z0-9][A-Z0-9][A-Z0-9][A-Z0-9] | (including CDL, state ID etc.) Starting September 2018, ID   number with 12 characters with “WDL” then 9 random (excluding A,E,I,O,U,Q and V)
Old WA DL |[A-Z][A-Z][A-Z][A-Z][A-Z][A-Z][A-Z*][A-Z0-9][A-Z0-9][A-Z0-9][A-Z0-9][A-Z0-9] | 6 capital letters, then 1   letter or symbol, then 5 letters or numbers

**Reply by will-saunders** [reply]
For more info on how Washington DL's work, check these pages, designed to help people recognize fake licenses
[ID Designs](https://www.dol.wa.gov/driverslicense/IDdesigns.html)  
[License designs](https://www.dol.wa.gov/driverslicense/licensedesigns.html)
Oh, and sorry it took a few days to close the loop on this, but there was Christmas etc. and yes we do have to check stuff thoroughly.
Pretty cool project!

**Reply by ammills01** [reply]
Dude you rock!  Thanks so much Will!  I'll get this new logic implemented as part of the DL REGEX and help improve false positives.  I'll make sure the REGEX correctly translates the logic you laid out.
